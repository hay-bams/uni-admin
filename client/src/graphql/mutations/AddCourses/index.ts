import { gql } from "@apollo/client";

export const ADD_COURSES = gql`
  mutation AddCourse($studentId: String!, $input: ID!) {
    addCourses(studentId: $studentId, input: $input) {
      id
    }
  }
`;