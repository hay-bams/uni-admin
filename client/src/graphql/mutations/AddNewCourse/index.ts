import { gql } from "@apollo/client";

export  const ADD_NEW_COURSE = gql`
  mutation AddNewCourse($input: NewCourseInput) {
    addNewCourse(input: $input) {
      id
    }
  }
`