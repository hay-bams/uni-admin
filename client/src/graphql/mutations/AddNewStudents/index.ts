import { gql } from "@apollo/client";

export  const ADD_NEW_STUDENT = gql`
  mutation AddNewStudent($input: NewStudentInput) {
    addNewStudent(input: $input) {
      id
    }
  }
`