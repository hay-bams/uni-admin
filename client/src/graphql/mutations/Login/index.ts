import { gql } from "@apollo/client";

export  const LOG_IN = gql`
  mutation Login($input: LoginInput) {
    login(input: $input) {
      id
      username
      token
      madeRequest
    }
  }
`