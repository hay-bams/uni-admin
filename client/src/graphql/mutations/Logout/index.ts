import { gql } from "@apollo/client";

export  const LOG_OUT = gql`
  mutation Logout {
    logout {
      id
      username
      madeRequest
    }
  }
`