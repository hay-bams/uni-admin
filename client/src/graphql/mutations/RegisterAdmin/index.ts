import { gql } from "@apollo/client";

export  const REGISTER_ADMIN = gql`
  mutation RegisterAdmin($input: RegisterInput) {
    register(input: $input) {
      id
      username
      madeRequest
    }
  }
`