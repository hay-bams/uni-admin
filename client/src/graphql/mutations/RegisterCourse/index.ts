import { gql } from "@apollo/client";

export const REGISTER_COURSE = gql`
  mutation RegisterCourse($studentId: String!, $input: ID!) {
    registerCourse(studentId: $studentId, input: $input) {
      id
    }
  }
`;