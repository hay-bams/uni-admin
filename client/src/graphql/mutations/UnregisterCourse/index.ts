import { gql } from "@apollo/client";

export const UNREGISTER_COURSE = gql`
  mutation UnregisterCourse($studentId: String!, $courseId: String!) {
    unregisterCourse(studentId: $studentId, courseId: $courseId) {
      id
    }
  }
`