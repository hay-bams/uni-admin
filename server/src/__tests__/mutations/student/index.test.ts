/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { COURSES_DATA, STUDENT_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import { TEST_REGISTER_COURSE, TEST_UNREGISTER_COURSE, TEST_ADD_NEW_STUDENT } from '../mutations';

describe('Student Mutation', () => {
  test('should add new student', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          insertOne: jest.fn(() => ({ ops: [STUDENT_DATA[0]] })),
          find: jest.fn(() => ({
            count: jest.fn(() => STUDENT_DATA),
          })),
        },
        courses: {
          find: jest.fn(() => ({
            toArray: jest.fn(() => COURSES_DATA),
          })),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_ADD_NEW_STUDENT,
      variables: {
        input:  {
          email: STUDENT_DATA[0].email,
          name: STUDENT_DATA[0].name,
          country: STUDENT_DATA[0].country
        },
      },
    });

    expect(res.data.addNewStudent).toHaveProperty('id');
    expect(res.data.addNewStudent).toHaveProperty('name');
    expect(res).toMatchSnapshot();
  })

  test('should register courses for a student', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          findOne: jest.fn(() => STUDENT_DATA[0]),
          findOneAndUpdate: jest.fn(() => ({
            value: STUDENT_DATA[0],
          })),
        },
        courses: {
          find: jest.fn(() => ({
            toArray: jest.fn(() => COURSES_DATA),
          })),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_REGISTER_COURSE,
      variables: {
        studentId: STUDENT_DATA[0]._id.toString(),
        input: COURSES_DATA[0]._id.toString(),
      },
    });

    expect(res.data.registerCourse).toHaveProperty('id');
    expect(res.data.registerCourse.courses.length).toBeGreaterThan(0);
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if student to be registered does not exist', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          findOne: jest.fn(() => null),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_REGISTER_COURSE,
      variables: {
        studentId: STUDENT_DATA[0]._id.toString(),
        input: COURSES_DATA[0]._id.toString(),
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if student course registration failed', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          findOne: jest.fn(() => STUDENT_DATA[0]),
          findOneAndUpdate: jest.fn(() => ({
            value: null,
          })),
        },
        courses: {
          find: jest.fn(() => ({
            toArray: jest.fn(() => COURSES_DATA),
          })),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_REGISTER_COURSE,
      variables: {
        studentId: STUDENT_DATA[0]._id.toString(),
        input: COURSES_DATA[0]._id.toString(),
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });

  test('should unregister courses for a student', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          findOneAndUpdate: jest.fn(() => ({
            value: STUDENT_DATA[0],
          })),
        },
        courses: {
          find: jest.fn(() => ({
            toArray: jest.fn(() => COURSES_DATA),
          })),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_UNREGISTER_COURSE,
      variables: {
        studentId: STUDENT_DATA[0]._id.toString(),
        courseId: COURSES_DATA[0]._id.toString(),
      },
    });

    expect(res.data.unregisterCourse).toHaveProperty('id');
    expect(res.data.unregisterCourse).toHaveProperty('courses');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if student course removal failed', async () => {
    const { mutate } = createTestServer({
      db: {
        students: {
          findOneAndUpdate: jest.fn(() => ({
            value: null,
          })),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_UNREGISTER_COURSE,
      variables: {
        studentId: STUDENT_DATA[0]._id.toString(),
        courseId: COURSES_DATA[0]._id.toString(),
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });
});
