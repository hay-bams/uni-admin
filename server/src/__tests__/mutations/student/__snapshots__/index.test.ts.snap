// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Mutation should register courses for a student 1`] = `
Object {
  "data": Object {
    "registerCourse": Object {
      "courses": Array [
        Object {
          "id": "5fdab398791e134a5466566f",
          "name": "Aerospace Engineerng",
        },
        Object {
          "id": "5fdab4b5eed1fb397feaca14",
          "name": "Fashion Technology",
        },
        Object {
          "id": "5fdab4bd78cc9321b31faf35",
          "name": "Marine Engineerng",
        },
        Object {
          "id": "5fdab4c4fe1aa062abc46096",
          "name": "Building, Contruction Management",
        },
        Object {
          "id": "5fdab4ccba6dcf525408412a",
          "name": "Agriculture",
        },
      ],
      "id": "5fda72349baf026a892dc1a0",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Student Mutation should throw an error if student course registration failed 1`] = `
Object {
  "data": Object {
    "registerCourse": null,
  },
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Courses could not be added],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Student Mutation should throw an error if student course removal failed 1`] = `
Object {
  "data": Object {
    "unregisterCourse": null,
  },
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Courses could not be removed],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Student Mutation should throw an error if student to be registered does not exist 1`] = `
Object {
  "data": Object {
    "registerCourse": null,
  },
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Student not found],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Student Mutation should unregister courses for a student 1`] = `
Object {
  "data": Object {
    "unregisterCourse": Object {
      "courses": Array [
        Object {
          "id": "5fdab398791e134a5466566f",
          "name": "Aerospace Engineerng",
        },
        Object {
          "id": "5fdab4b5eed1fb397feaca14",
          "name": "Fashion Technology",
        },
        Object {
          "id": "5fdab4bd78cc9321b31faf35",
          "name": "Marine Engineerng",
        },
        Object {
          "id": "5fdab4c4fe1aa062abc46096",
          "name": "Building, Contruction Management",
        },
        Object {
          "id": "5fdab4ccba6dcf525408412a",
          "name": "Agriculture",
        },
      ],
      "id": "5fda72349baf026a892dc1a0",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
